be/src/exec/workgroup/work_group.cpp

```diff
@@ -89,12 +97,206 @@ void WorkGroupManager::remove_workgroup(int wg_id) {
     }
 }
 
-WorkGroupPtr WorkGroupManager::pick_next_wg_for_cpu() {
-    return _wg_cpu_queue.pick_next();
+bool WorkGroup::try_offer_io_task(const PriorityThreadPool::Task& task) {
+    return _io_work_queue.try_put(task);
+}
+
+void WorkGroup::pick_and_run_io_task() {
+    PriorityThreadPool::Task task;
+    if (_io_work_queue.blocking_get(&task)) {
+        task.work_function();
+    }
+}
+
+// shuold be call when read chunk_num from disk
+void WorkGroup::increase_chunk_num(int32_t chunk_num) {
+    _cur_hold_total_chunk_num += chunk_num;
+    _increase_chunk_num_period += chunk_num;
+}
+
+void WorkGroup::decrease_chunk_num(int32_t chunk_num) {
+    _cur_hold_total_chunk_num -= chunk_num;
+    _decrease_chunk_num_period -= chunk_num;
+}
+
+double WorkGroup::get_expect_factor() const {
+    return _expect_factor;
+}
+
+double WorkGroup::get_diff_factor() const {
+    return _diff_factor;
+}
+
+double WorkGroup::get_select_factor() const {
+    return _select_factor;
+}
+
+void WorkGroup::update_select_factor(double value) {
+    _select_factor += value;
+}
+
+double WorkGroup::get_cur_select_factor() const {
+    return _cur_select_factor;
+}
+
+void WorkGroup::update_cur_select_factor(double value) {
+    _cur_select_factor += value;
+}
+
+void WorkGroup::estimate_trend_factor_period() {
+    // maybe should add a guard
+
+    // Exapmle
+    // during a schedule period, actually this WorkGroup consume 70 chunk and use cpu resource percent 80
+    // but this WorkGroup cpu limit to percent 80
+    // meanwhile, WorkGroup read 200 chunk from io_threads, so we calculate some factor bellow
+    // decrease_factor = 70 / 0.8 * (0.7 / 0.8)
+    // increase_factor = 200 / 0.7
+    // _expect_factor mean we except io resource ratio of this WorkGroup to suit to 70 percent cpu limit
+    // so _expect_factor = decrease_factor / increase_factor * 70
+
+    double decrease_factor =
+            _decrease_chunk_num_period / _cpu_actual_use_ratio * _cpu_expect_use_ratio / _cpu_actual_use_ratio;
+    double increase_factor = _increase_chunk_num_period / _cpu_expect_use_ratio;
+
+    _expect_factor = decrease_factor / increase_factor * _cpu_expect_use_ratio;
+
+    // diff_factor mean diff between actually need io resource percent and limit percent of cpu
+    // if it is nagative, it mean resouse less and need more
+    // if it is positive, it mean resouse is enough and can be reduce
+    _diff_factor = _cpu_expect_use_ratio - _expect_factor;
+
+    // it only use for this period to calculate factors
+    // over calculate, it must be reset to zero for next period
+    _decrease_chunk_num_period = 0;
+    _increase_chunk_num_period = 0;
+}
+
+void WorkGroupManager::adjust_weight_if_need() {
+    if (--_cur_schedule_num >= 0) {
+        return;
+    }
+
+    bool expect = false;
+    if (!_is_adjusted.compare_exchange_strong(expect, true)) {
```

Comment:
Similar as the following `_is_scheduled.compare_exchange_strong`. 

Assume that two threads find `--_cur_schedule_num < 0`, and check CAS here.
If thread-1 has already finished adjusted before thread-2 checks CAS, then thread-2 will adjust again. Is it ok for this situation? Or should we use an atomic `_adjusted_sequence_num` to avoid this?